cql3/query_processor.cc

```diff
@@ -518,6 +524,21 @@ query_processor::execute_prepared_without_checking_exception_message(
     });
 }
 
+static
+future<::shared_ptr<result_message>>
+dispatch_statement_execution(query_processor& qp, const cql_statement& statement, service::query_state& query_state, const query_options& options) {
+    if (qp.has_group0_tables() && raft::group0_tables::is_group0_table_statement(statement)) {
+        if (this_shard_id() != 0) {
+            co_return qp.bounce_to_shard(0, {});
+        }
+
+        co_await raft::group0_tables::execute(qp.get_group0_client(), statement);
+        co_return ::shared_ptr<result_message>{};
+    } else [[likely]] {
+        co_return co_await statement.execute_without_checking_exception_message(qp, query_state, options);
+    }
+}
```

Comment:
This pessimizes the common case by introducing a coroutine (=allocation) even when you don't need it. This should be a non-coroutine with just an if, calling the two branches (which may or may not be coroutines).

But, I don't think this is the correct place. We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing. This way we only have one dispatch point in the fast path, not two.

Reply:
> This way we only have one dispatch point in the fast path, not two.

How so? We would dispatch on `prepare` to decide which of the subclasses to create, then a dispatch on `execute_without_checking_exception_message` - a virtual call. That's two dispatches.

Reply:
prepare() is amortized over all of the uses of the statement. In a normal execution, it is not called.

Reply:
>  We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing.

`cql_statement` has a whole tree of subclasses: `modification_statement`, `select_statement`, `batch_statement` and so on.

Do you suggest that we should have a mirror tree for strongly consistent tables? Like `strongly_consistent_modification_statement`, `strongly_consistent_select_statement` etc.?

Please elaborate.

If this is your proposal, then I think a better approach would be to simply have a `compile` virtual function in the existing hierarchy.

Reply:
> > We should have prepare() create new subclass of cql_statement with its own execute_without_checking_exception_message() which can then do its thing.
> 
> `cql_statement` has a whole tree of subclasses: `modification_statement`, `select_statement`, `batch_statement` and so on.

In fact, there are several variations of select_statement, each corresponding to a different query plan (regular select, indexed, parallelized).

> 
> Do you suggest that we should have a mirror tree for strongly consistent tables? Like `strongly_consistent_modification_statement`, `strongly_consistent_select_statement` etc.?
> 
> Please elaborate.

Yes. They can share a lot of their internals. For example, SELECT will forward to the leader, which will then just reuse the eventually consistent select (perhaps with a read barrier too).

> 
> If this is your proposal, then I think a better approach would be to simply have a `compile` virtual function in the existing hierarchy.

compile() exists, it's called prepare().

Reply:
> compile() exists, it's called prepare().

`prepare` creates an object in an inheritance hierarchy. We need a `compile` which produces a (de)serializable (to/from bytes) data structure which we can put inside a Raft command. Even if we have a separate inheritance hierarchy, there will be an internal `compile` somewhere.

Reply:
prepare() can create an object with the compiled query (CQL or otherwise) who's execute_with_some_long_other_things() sends to raft. The compilation must occur ahead of time, we cannot compile each time a query is executed.

Reply:
> prepare() can create an object with the compiled query (CQL or otherwise) who's execute_with_some_long_other_things() sends to raft.

Did I understand correctly that you propose to add new subclass of `cql_statement` whose `execute_without_checking_exception_message` method will
1. compile `query`
2. send `query` to raft

and leave executing queries in state machine as it is now?

Reply:
I don't think we want to compile in `execute_without_checking_exception_message`. We want to compile ahead of time, in `prepare`, then `execute_...` will only send to raft.

Reply:
Yes. Execution should attach run-time parameters to a handle to a pre-compiled query and send it along.

Calling something compile() is a sure way to get it rejected from functions called execute().

Reply:
Moved compilation to `prepare()`. Executing pre-compiled queries with parameters will be added in the future.